Add resource throw NPE when use database with upper database name

## Bug Report

### Which version of ShardingSphere did you use?

https://github.com/apache/shardingsphere/commit/4f15b70e828888c0d9880f1e32c623c272b73840

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Add resources successfully when use database with upper database name.

### Actual behavior

```
org.apache.zookeeper.ClientCnxn$SessionTimeoutException: Client session timed out, have not heard from server in 58213ms for session id 0x100000046bc0040
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1230)
[ERROR] 2022-07-13 18:27:28.396 [ShardingSphere-Command-0] o.a.s.p.b.t.d.r.r.AddResourceBackendHandler - Add resource failed
java.lang.NullPointerException: null
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
	at org.apache.shardingsphere.infra.federation.optimizer.context.OptimizerContext.alterDatabase(OptimizerContext.java:73)
	at org.apache.shardingsphere.mode.manager.ContextManager.updateResources(ContextManager.java:209)
	at org.apache.shardingsphere.proxy.backend.text.distsql.rdl.resource.AddResourceBackendHandler.execute(AddResourceBackendHandler.java:63)
	at org.apache.shardingsphere.proxy.backend.text.distsql.rdl.resource.AddResourceBackendHandler.execute(AddResourceBackendHandler.java:44)
	at org.apache.shardingsphere.proxy.backend.text.DatabaseRequiredBackendHandler.execute(DatabaseRequiredBackendHandler.java:51)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:95)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:108)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 2022-07-13 18:27:28.399 [ShardingSphere-Command-0] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
org.apache.shardingsphere.infra.distsql.exception.resource.InvalidResourcesException: Can not process invalid resources, error messages are: [null].
	at org.apache.shardingsphere.proxy.backend.text.distsql.rdl.resource.AddResourceBackendHandler.execute(AddResourceBackendHandler.java:68)
	at org.apache.shardingsphere.proxy.backend.text.distsql.rdl.resource.AddResourceBackendHandler.execute(AddResourceBackendHandler.java:44)
	at org.apache.shardingsphere.proxy.backend.text.DatabaseRequiredBackendHandler.execute(DatabaseRequiredBackendHandler.java:51)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:95)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:108)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
[INFO ] 2022-07-13 18:27:28.489 [Quartz Shutdown-Hook _finished_check] o.a.s.e.l.i.s.JobShutdownHookPlugin - Shutting down Quartz... _finished_check
[INFO ] 2022-07-13 18:27:28.489 [Quartz Shutdown-Hook _finished_check] org.quartz.core.QuartzScheduler - Scheduler _finished_check_$_NON_CLUSTERED shutting down.
[INFO ] 2022-07-13 18:27:28.489 [Quartz Shutdown-Hook _finished_check] org.quartz.core.QuartzScheduler - Scheduler _finished_check_$_NON_CLUSTERED paused.
[INFO ] 2022-07-13 18:27:28.501 [_finished_check_Worker-1-EventThread] o.a.c.f.state.ConnectionStateManager - State change: SUSPENDED
[INFO ] 2022-07-13 18:27:28.501 [main-EventThread] o.a.c.f.state.ConnectionStateManager - State change: SUSPENDED
[INFO ] 2022-07-13 18:27:28.502 [main-EventThread] o.a.c.f.state.ConnectionStateManager - State change: SUSPENDED
```

### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/context/OptimizerContext.java
+++ b/shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/main/java/org/apache/shardingsphere/infra/federation/optimizer/context/OptimizerContext.java
@@ -56,10 +56,10 @@ public final class OptimizerContext {
      * @param protocolType protocol database type
      */
     public void addDatabase(final String databaseName, final DatabaseType protocolType) {
-        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData(databaseName, Collections.emptyMap());
-        federationMetaData.getDatabases().put(databaseName, federationDatabaseMetaData);
-        parserContexts.put(databaseName, OptimizerParserContextFactory.create(protocolType));
-        plannerContexts.put(databaseName, OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
+        FederationDatabaseMetaData federationDatabaseMetaData = new FederationDatabaseMetaData(databaseName.toLowerCase(), Collections.emptyMap());
+        federationMetaData.getDatabases().put(databaseName.toLowerCase(), federationDatabaseMetaData);
+        parserContexts.put(databaseName.toLowerCase(), OptimizerParserContextFactory.create(protocolType));
+        plannerContexts.put(databaseName.toLowerCase(), OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
     }
     
     /**
@@ -69,9 +69,10 @@ public final class OptimizerContext {
      * @param globalRuleMetaData global rule meta data
      */
     public void alterDatabase(final ShardingSphereDatabase database, final ShardingSphereRuleMetaData globalRuleMetaData) {
-        OptimizerContext toBeAlteredOptimizerContext = OptimizerContextFactory.create(Collections.singletonMap(database.getName(), database), globalRuleMetaData);
-        federationMetaData.getDatabases().put(database.getName(), toBeAlteredOptimizerContext.getFederationMetaData().getDatabases().get(database.getName()));
-        plannerContexts.put(database.getName(), toBeAlteredOptimizerContext.getPlannerContexts().get(database.getName()));
+        String databaseName = database.getName().toLowerCase();
+        OptimizerContext toBeAlteredOptimizerContext = OptimizerContextFactory.create(Collections.singletonMap(databaseName, database), globalRuleMetaData);
+        federationMetaData.getDatabases().put(databaseName, toBeAlteredOptimizerContext.getFederationMetaData().getDatabases().get(databaseName));
+        plannerContexts.put(databaseName, toBeAlteredOptimizerContext.getPlannerContexts().get(databaseName));
     }
     
     /**
@@ -80,9 +81,9 @@ public final class OptimizerContext {
      * @param databaseName database name
      */
     public void dropDatabase(final String databaseName) {
-        federationMetaData.getDatabases().remove(databaseName);
-        parserContexts.remove(databaseName);
-        plannerContexts.remove(databaseName);
+        federationMetaData.getDatabases().remove(databaseName.toLowerCase());
+        parserContexts.remove(databaseName.toLowerCase());
+        plannerContexts.remove(databaseName.toLowerCase());
     }
     
     /**
@@ -92,9 +93,9 @@ public final class OptimizerContext {
      * @param schemaName schema name
      */
     public void addSchema(final String databaseName, final String schemaName) {
-        federationMetaData.getDatabases().get(databaseName).putTable(schemaName, new ShardingSphereTable());
+        federationMetaData.getDatabases().get(databaseName.toLowerCase()).putTable(schemaName, new ShardingSphereTable());
         // TODO add schema only
-        plannerContexts.put(databaseName, OptimizerPlannerContextFactory.create(federationMetaData.getDatabases().get(databaseName)));
+        plannerContexts.put(databaseName.toLowerCase(), OptimizerPlannerContextFactory.create(federationMetaData.getDatabases().get(databaseName.toLowerCase())));
     }
     
     /**
@@ -105,9 +106,9 @@ public final class OptimizerContext {
      * @param toBeChangedTable to be changed table
      */
     public void alterTable(final String databaseName, final String schemaName, final ShardingSphereTable toBeChangedTable) {
-        FederationDatabaseMetaData federationDatabaseMetaData = federationMetaData.getDatabases().get(databaseName);
-        federationDatabaseMetaData.putTable(schemaName, toBeChangedTable);
-        plannerContexts.put(databaseName, OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
+        FederationDatabaseMetaData federationDatabaseMetaData = federationMetaData.getDatabases().get(databaseName.toLowerCase());
+        federationDatabaseMetaData.putTable(schemaName.toLowerCase(), toBeChangedTable);
+        plannerContexts.put(databaseName.toLowerCase(), OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
     }
     
     /**
@@ -118,8 +119,8 @@ public final class OptimizerContext {
      * @param toBeDeletedTableName to be deleted table name
      */
     public void dropTable(final String databaseName, final String schemaName, final String toBeDeletedTableName) {
-        FederationDatabaseMetaData federationDatabaseMetaData = federationMetaData.getDatabases().get(databaseName);
+        FederationDatabaseMetaData federationDatabaseMetaData = federationMetaData.getDatabases().get(databaseName.toLowerCase());
         federationDatabaseMetaData.removeTableMetadata(schemaName, toBeDeletedTableName);
-        plannerContexts.put(databaseName, OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
+        plannerContexts.put(databaseName.toLowerCase(), OptimizerPlannerContextFactory.create(federationDatabaseMetaData));
     }
 }

--- a/shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/context/OptimizerContextTest.java
+++ b/shardingsphere-infra/shardingsphere-infra-federation/shardingsphere-infra-federation-optimizer/src/test/java/org/apache/shardingsphere/infra/federation/optimizer/context/OptimizerContextTest.java
@@ -42,14 +42,14 @@ public final class OptimizerContextTest {
     @Test
     public void assertDropDatabase() {
         OptimizerContext optimizerContext = createOptimizerContext();
-        optimizerContext.dropDatabase("foo_db");
+        optimizerContext.dropDatabase("FOO_DB");
         assertFalse(optimizerContext.getFederationMetaData().getDatabases().containsKey("foo_db"));
     }
     
     @Test
     public void assertAlterTable() {
         OptimizerContext optimizerContext = createOptimizerContext();
-        optimizerContext.alterTable("foo_db", "foo_schema", createTable("bar_col"));
+        optimizerContext.alterTable("FOO_DB", "foo_schema", createTable("bar_col"));
         Optional<FederationSchemaMetaData> schemaMetaData = optimizerContext.getFederationMetaData().getDatabases().get("foo_db").getSchemaMetadata("foo_schema");
         assertTrue(schemaMetaData.isPresent());
         assertFalse(schemaMetaData.get().getTables().get("foo_tbl").getColumnNames().contains("foo_col"));
@@ -60,7 +60,7 @@ public final class OptimizerContextTest {
     public void assertDropTable() {
         OptimizerContext optimizerContext = createOptimizerContext();
         OptimizerPlannerContext beforeDroppedPlannerContext = optimizerContext.getPlannerContexts().get("foo_db");
-        optimizerContext.dropTable("foo_db", "foo_schema", "foo_tbl");
+        optimizerContext.dropTable("FOO_DB", "foo_schema", "foo_tbl");
         assertThat(beforeDroppedPlannerContext, not(optimizerContext.getPlannerContexts().get("foo_db")));
         Optional<FederationSchemaMetaData> schemaMetadata = optimizerContext.getFederationMetaData().getDatabases().get("foo_db").getSchemaMetadata("foo_schema");
         assertTrue(schemaMetadata.isPresent());
@@ -70,7 +70,7 @@ public final class OptimizerContextTest {
     @Test
     public void assertAddDatabase() {
         OptimizerContext optimizerContext = createOptimizerContext();
-        optimizerContext.addDatabase("bar_db", new H2DatabaseType());
+        optimizerContext.addDatabase("BAR_DB", new H2DatabaseType());
         assertTrue(optimizerContext.getFederationMetaData().getDatabases().containsKey("bar_db"));
         assertTrue(optimizerContext.getParserContexts().containsKey("bar_db"));
         assertTrue(optimizerContext.getPlannerContexts().containsKey("bar_db"));
@@ -85,6 +85,15 @@ public final class OptimizerContextTest {
         Optional<FederationSchemaMetaData> schemaMetadata = optimizerContext.getFederationMetaData().getDatabases().get("foo_db").getSchemaMetadata("foo_schema");
         assertTrue(schemaMetadata.isPresent());
         assertTrue(schemaMetadata.get().getTables().containsKey("bar_tbl"));
+    }
+    
+    @Test
+    public void assertAddSchema() {
+        OptimizerContext optimizerContext = createOptimizerContext();
+        optimizerContext.addSchema("FOO_DB", "foo_schema");
+        assertTrue(optimizerContext.getFederationMetaData().getDatabases().get("foo_db").getSchemas().containsKey("foo_schema"));
+        assertTrue(optimizerContext.getPlannerContexts().get("foo_db").getConverters().containsKey("foo_schema"));
+        assertTrue(optimizerContext.getPlannerContexts().get("foo_db").getValidators().containsKey("foo_schema"));
     }
     
     private OptimizerContext createOptimizerContext() {

